pkg/controller/storagecluster/initialization_reconciler.go

```diff
@@ -297,6 +305,94 @@ func (r *ReconcileStorageCluster) newStorageClasses(initData *ocsv1.StorageClust
+func (r *ReconcileStorageCluster) newRBDSnapshotClass(initData *ocsv1.StorageCluster) *snapapi.VolumeSnapshotClass {
+	retSC := &snapapi.VolumeSnapshotClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: generateNameForSnapshotClass(initData, "rbd"),
+		},
+		Driver: generateNameForSnapshotClassDriver(initData, "rbd"),
+		Parameters: map[string]string{
+			"clusterID":                initData.Namespace,
+			snapshotterSecretName:      "rook-csi-rbd-provisioner",
+			snapshotterSecretNamespace: initData.Namespace,
+		},
+		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
+	}
+	return retSC
+}
+
+// newCephFsSnapshotClass returns a new CephFs snapshotter backed SnapshotClass
+func (r *ReconcileStorageCluster) newCephFsSnapshotClass(initData *ocsv1.StorageCluster) *snapapi.VolumeSnapshotClass {
+	retSC := &snapapi.VolumeSnapshotClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: generateNameForSnapshotClass(initData, "cephfs"),
+		},
+		Driver: generateNameForSnapshotClassDriver(initData, "cephfs"),
+		Parameters: map[string]string{
+			"clusterID":                initData.Namespace,
+			snapshotterNamePrefix:      fmt.Sprintf("%s-%s-", initData.Name, initData.Namespace),
+			snapshotterSecretName:      "csi-cephfs-secret",
+			snapshotterSecretNamespace: initData.Namespace,
+		},
+		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
+	}
+	return retSC
+}
+
+func (r *ReconcileStorageCluster) newSnapshotClasses(instance *ocsv1.StorageCluster) []*snapapi.VolumeSnapshotClass {
+	scs := []*snapapi.VolumeSnapshotClass{
+		r.newRBDSnapshotClass(instance),
+		r.newCephFsSnapshotClass(instance),
+	}
+	return scs
+}
+
+func (r *ReconcileStorageCluster) createSnapshotClasses(vscs []*snapapi.VolumeSnapshotClass, reqLogger logr.Logger) error {
+	for _, vsc := range vscs {
+		existing := &snapapi.VolumeSnapshotClass{}
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: vsc.Name, Namespace: vsc.Namespace}, existing)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Since the SnapshotClass is not found, we will create a new one
+				reqLogger.Info(fmt.Sprintf("Creating SnapshotClass %s", vsc.Name))
+				err = r.client.Create(context.TODO(), vsc)
+			} // on any (other) error, including (above) resource creation error, will be dealt below
+		}
+		// if there is *still* an error, we should report it
+		if err != nil {
+			return err
+		}
+		if existing.DeletionTimestamp != nil {
+			return fmt.Errorf("failed to restore snapshotclass  %s because it is marked for deletion", existing.Name)
+		}
+		// if there is a mis-match in the parameters of existing vs created resources,
+		if !reflect.DeepEqual(vsc.Parameters, existing.Parameters) {
+			// we have to update the existing SnapshotClass
+			reqLogger.Info(fmt.Sprintf("SnapshotClass %s needs to be updated", existing.Name))
+			existing.ObjectMeta.OwnerReferences = vsc.ObjectMeta.OwnerReferences
+			vsc.ObjectMeta = existing.ObjectMeta
```

Comment:
if there is a mismatch , are we falling back to `existing` instead of `vsc` ?

Reply:
This is the case of update. It's just copying metadata of existing resource and adding it to the new resource for updation. Else it would throw an invalid object error during updates.